lib/apiservers/engine/backends/container.go

```diff
@@ -537,59 +585,121 @@ func requestHostPort(proto string) (int, error) {
-	}
+// unrollPortMap processes config for mapping/unmapping ports e.g. from hostconfig.PortBindings
+func unrollPortMap(portMap nat.PortMap) ([]*portMapping, error) {
+	var portMaps []*portMapping
+	for i, pb := range portMap {
 
-	_, bindings, err := nat.ParsePortSpecs(endpoint.Ports)
-	if err != nil {
-		return err
-	}
-	for p := range bindings {
-		proto, port := nat.SplitProtoPort(string(p))
-		var nport nat.Port
+		proto, port := nat.SplitProtoPort(string(i))
 		nport, err := nat.NewPort(proto, port)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		pbs, ok := hostconfig.PortBindings[nport]
-		if !ok {
-			continue
-		}
-
-		for i := range pbs {
+		// iterate over all the ports in pb []nat.PortBinding
+		for _, p := range pb {
 			var hostPort int
-			if pbs[i].HostPort == "" {
+			var hPort string
+			if p.HostPort == "" {
 				// use a random port since no host port is specified
 				hostPort, err = requestHostPort(proto)
 				if err != nil {
 					log.Errorf("could not find available port on host")
-					return err
+					return nil, err
 				}
 				// update the hostconfig
-				pbs[i].HostPort = strconv.Itoa(hostPort)
+				p.HostPort = strconv.Itoa(hostPort)
 
 			} else {
-				hostPort, err = strconv.Atoi(pbs[i].HostPort)
+				hostPort, err = strconv.Atoi(p.HostPort)
 				if err != nil {
-					return err
+					return nil, err
 				}
 			}
+			hPort = strconv.Itoa(hostPort)
+			portMaps = append(portMaps, &portMapping{
+				intHostPort: hostPort,
+				strHostPort: hPort,
+				portProto:   nport,
+			})
+		}
+	}
+	return portMaps, nil
+}
 
-			if err = portMapper.MapPort(op, nil, hostPort, nport.Proto(), containerIP.String(), nport.Int(), clientIfaceName, bridgeIfaceName); err != nil {
-				return err
-			}
+// mapPorts maps ports defined in hostconfig for containerID
+func (c *Container) mapPorts(hostconfig *containertypes.HostConfig, endpoint *models.EndpointConfig, containerID string) error {
+	log.Debugf("mapPorts for %q: %v", containerID, hostconfig.PortBindings)
+
+	if len(hostconfig.PortBindings) == 0 {
+		return nil
+	}
+	if endpoint == nil {
+		return nil
+	}
+
+	var containerIP net.IP
+	containerIP = net.ParseIP(endpoint.Address)
+	if containerIP == nil {
+		return fmt.Errorf("invalid endpoint address %s", endpoint.Address)
+	}
+
+	portMap, err := unrollPortMap(hostconfig.PortBindings)
+	if err != nil {
+		return err
+	}
+
+	cbpLock.Lock()
+	defer cbpLock.Unlock()
+	for _, p := range portMap {
+		if err = portMapper.MapPort(nil, p.intHostPort, p.portProto.Proto(), containerIP.String(), p.portProto.Int(), clientIfaceName, bridgeIfaceName); err != nil {
+			return err
 		}
+
+		// update mapped ports
+		containerByPort[p.strHostPort] = containerID
+		log.Debugf("mapped port %s for container %s", p.strHostPort, containerID)
 	}
+	return nil
+}
 
+// unmapPorts unmaps ports defined in hostconfig for containerID
+func (c *Container) unmapPorts(hostconfig *containertypes.HostConfig, containerID string) error {
```

Comment:
`conatinerID` is not used by the function at all other than for debugging statements. Can we take it out?